src/fhirtypes/ElementDefinition.ts

```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
+   * @param {string} target - the target to find a matching type for
+   * @param {ResolveFn} resolve - a function that can resolve a type to a StructureDefinition instance
+   * @returns {ElementDefinitionType} the element's type that matches the target
+   * @throws {TypeNotFoundError} when the target's definition cannot be found
+   * @throws {InvalidTypeError} when the target doesn't match any existing types
+   */
+  private getTargetType(target: string, resolve: ResolveFn): ElementDefinitionType {
+    let targetType: ElementDefinitionType;
+    if (target) {
+      const targetSD = resolve(target);
+      if (targetSD == null) {
+        throw new TypeNotFoundError(target);
+      }
+      // Try to match on types by an exact match on the code (applies to resources),
+      // the profiles (applies to profiles), or targetProfiles (applies to references).
+      // Clone it since we will filter out the non-target profiles/targetProfiles.
+      targetType = cloneDeep(
+        this.type.find(
+          t =>
+            t.code === targetSD.id ||
+            t.profile?.includes(targetSD.url) ||
+            t.targetProfile?.includes(targetSD.url)
+        )
+      );
+
+      if (!targetType) {
+        throw new InvalidTypeError(target, this.type);
+      }
+
+      // Re-assign the targetProfiles/profiles as appopriate to remove non-targets
+      if (targetType.profile?.includes(targetSD.url)) {
+        targetType.profile = [targetSD.url];
+      } else if (targetType.targetProfile?.includes(targetSD.url)) {
+        targetType.targetProfile = [targetSD.url];
+      }
+    }
+    return targetType;
+  }
+
+  /**
+   * Given an input type (the constraint) and a set of target types (the things to potentially
+   * constrain), find the match and return information about it.
+   * @param {{ type: string; isReference?: boolean }} type - the constrained types, identified by
+   *   id/type/url string and an optional reference flag (defaults false)
+   * @param {ElementDefinitionType[]} targetTypes - the element types that the constrained type
+   *   can be potentially applied to
+   * @param {ResolveFn} resolve - a function that can resolve a type to a StructureDefinition instance
+   * @param {string} [target] - a specific target type to constrain.  If supplied, will attempt to
+   *   constrain only that type without affecting other types (in a choice or reference to a choice).
+   * @returns {ElementTypeMatchInfo} the information about the match
+   * @throws {TypeNotFoundError} when the type's definition cannot be found
+   * @throws {InvalidTypeError} when the type doesn't match any of the targetTypes
+   */
+  private findTypeMatch(
+    type: { type: string; isReference?: boolean },
+    targetTypes: ElementDefinitionType[],
+    resolve: ResolveFn
+  ): ElementTypeMatchInfo {
+    let matchedType: ElementDefinitionType;
+
+    // Get the lineage (type hierarchy) so we can walk up it when attempting to match
```

Comment:
Lineage is a very nice term for this. I like it.

Reply:
Thanks.  I felt a bit like I was a historian studying royalty, but it was fun nonetheless.